python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -303,70 +388,92 @@ def __reduce__(self):
+                    "Expected torch.Tensors with shapes and dtypes: "
+                    "[" + ", ".join(expected_str) + "],\n"
+                    "found: [" + ", ".join(metadata_str) + "]"
+                )
+            # The receiver has already determined the shape and dtype of the
+            # tensors from a previous send, so no need to send the metadata
+            # again.
+            return None
+
+        if self._static_shape:
+            # The shape and dtype is static. This is the first send op and
+            # afterwards, a ValueError will be thrown if the sent tensors do
+            # not match this metadata.
+            self._static_tensor_metadata = metadata_list
+        return metadata_list
 
     def write(
         self,
-        tensors: Union["torch.Tensor", List["torch.Tensor"], Exception],
+        tensors: List["torch.Tensor"],
         timeout: Optional[float] = None,
     ):
-        if isinstance(tensors, ray.exceptions.RayTaskError):
-            # TODO(swang): Write exceptions to the meta channel if it is
-            # available.
-            raise tensors
-
-        if isinstance(tensors, list):
-            meta_list = []
-            for tensor in tensors:
-                meta_list.append(self._get_tensor_meta(tensor))
-            if self.has_static_type():
-                # Make sure that there is exactly one tensor to send, and its
-                # metadata should have matched the static type.
-                if meta_list != [None]:
-                    raise ValueError(
-                        "DAGNode annotated with "
-                        "TorchTensorType(_shape=shape, _dtype=dtype))` can return at "
-                        "most one tensor with the declared `_shape` and `_dtype`. "
-                        "Use TorchTensorType() if value contains more than one "
-                        "tensor or tensor of dynamic size."
-                    )
-            else:
-                self._meta_channel.write(meta_list)
-        else:
-            meta = self._get_tensor_meta(tensors)
-            if meta is not None:
-                self._meta_channel.write(meta)
-            tensors = [tensors]
+        """
+        Write a list of tensors via NCCL:
+
+        1) Send the tensor metadata, i.e. the shape and dtypes of all tensors
+        via the shared-memory metadata channel.
+        2) Send the tensor data via NCCL.
+
+        If static_shape=True was set, then we only perform step (1) on the
+        first message. The reader is expected to reuse the sent metadata for
+        subsequent messages.
+        """
+        import torch
+
+        for tensor in tensors:
+            assert isinstance(
+                tensor, torch.Tensor
+            ), f"{tensor} must be instance of torch.Tensor"
+
+        # Send the tensors metadata so that the receiver knows what buffers to
+        # allocate.
+        metadata = self._get_send_tensors_metadata(tensors)
```

Comment:
should we check if the metadata and tensor matches? Or is it checked in the reader side?

Reply:
It is already checked on the writer in `_get_send_tensors_metadata`